src/msolve/hilbert.c

```diff
@@ -2900,6 +2984,225 @@ static inline void build_matrixn_from_bs_trace_application(sp_matfglm_t *matrix,
+      evi[i-2]    =   i;
+    }
+  }
+  bs_t *tbr;
+  if (count_not_lm) {
+    md_t *md = copy_meta_data(st,fc);
+    tbr = initialize_basis(md);
+    exp_t *mul = (exp_t *)calloc(ht->evl, sizeof(exp_t));
+    tbr->ht = ht;
+    /* reduction */
+    import_input_data(tbr, md, 0, count_not_lm, lens_extra_nf, exps_extra_nf,
+		      (void *)cfs_extra_nf, NULL);
+    tbr->ld = tbr->lml  =  count_not_lm;
+    for (long k = 0; k < count_not_lm; ++k) {
+      tbr->lmps[k]  = k; /* fix input element in bs */
+    }
+    int32_t err = 0;
+    tbr = core_nf(tbr, md, mul, bs, &err);
+    if (err) {
+      printf("Problem with normalform, stopped computation.\n");
+      exit(1);
+    }
+    free(mul);
+    free(md);
+  }
+  long len0 = matrix->nrows;
+  long len_xn = len0-count_not_lm; //get_div_xn(bexp_lm, bs->lml, nv, div_xn);
+
+  matrix->charac = fc;
+  long len1 = dquot * len0;
+  long len2 = dquot - len0;
+
+  for(long i = 0; i < len1; i++){
+    matrix->dense_mat[i] = 0;
+  }
+  for(long i = 0; i < len2; i++){
+    matrix->triv_idx[i] = 0;
+  }
+  for(long i = 0; i < len2; i++){
+    matrix->triv_pos[i] = 0;
+  }
+  for(long i = 0; i < len0; i++){
+    matrix->dense_idx[i] = 0;
+  }
+  for(long i = 0; i < len0; i++){
+    matrix->dst[i] = 0;
+  }
+
+  long pos = 0, k = 0;
+  for(long i = 0; i < bs->lml; i++){
+    long len = bs->hm[bs->lmps[i]][LENGTH];
+    if(i==div_xn[k]){
+      len_gb_xn[k]=len;
+      start_cf_gb_xn[k]=pos;
+      pos+=len;
+      k++;
+    }
+    else{
+      pos+=len;
+    }
+  }
+
+  long l_triv = 0;
+  long l_dens = 0;
+  long nrows = 0;
+  long count = 0;
+  long count_nf = 0;
+  
+  for(long i = 0; i < dquot; i++){
+    long pos = -1;
+    int32_t *exp = lmb + (i * nv);
+#if DEBUGBUILDMATRIX > 0
+    display_monomial_full(stderr, nv, NULL, 0, exp);
+#endif
+    if(member_xxn(exp, lmb + (i * nv), dquot - i, &pos, nv)){
+#if DEBUGBUILDMATRIX > 0
+      fprintf(stderr, " => remains in monomial basis\n");
+#endif
+
+      matrix->triv_idx[l_triv] = i;
+      matrix->triv_pos[l_triv] = pos + i;
+
+      l_triv++;
+    }
+    else{
+
+#if DEBUGBUILDMATRIX > 0
+      fprintf(stderr, " => does NOT remain in monomial basis\n");
+#endif
+      matrix->dense_idx[l_dens] = i;
+      l_dens++;
+      if(is_equal_exponent_xxn(exp, bexp_lm+(div_xn[count])*nv, nv)){
+        copy_poly_in_matrix_from_bs(matrix, nrows, bs, ht, //bcf, bexp, blen,
+                                    div_xn[count], len_gb_xn[count],
+                                    start_cf_gb_xn[count], len_gb_xn[count], lmb,
+                                    nv, fc);
+        nrows++;
+        count++;
+        if(len_xn < count && i < dquot){
+          fprintf(stderr, "One should not arrive here (build_matrix with trace)\n");
+          free(matrix->dense_mat);
+          free(matrix->dense_idx);
+          free(matrix->triv_idx);
+          free(matrix->triv_pos);
+          free(matrix->dst);
+          free(matrix);
+
+	  free(tbr);
```

Comment:
```suggestion
	  free_basis_without_hash_table(&tbr);
```
This should be applied to all instances of `free(tbr)` in your new code. This removes at least some memory leaks,  further looking into it.

Reply:
Thanks, this is changed!